src/model/model.py

```diff
@@ -160,6 +160,105 @@ def apply_operation(self, arr, elements=[], operation='add'):
             arr = function(arr, element, dtype=np.float64)
         return arr
 
+    def apply_formula(self, elements, formula, initialize=True, id=0):
+        """
+        This function calculate complexe operations between float/int and images
+        it handles parenthesis grouping and operator priorities.
+        And it calls "apply_operation" to do pear operations recursively.
+
+        Parameters
+        ----------
+        elements: dict
+            this dictionary contains data (image or float/int...)
+        formula: str
+
+        Return
+        ------
+        result: np.ndarray or int/float
+            the result type depends on the formula.
+
+        Example
+        -------
+        formula = "( 3 x [key_1] + 4 x [key_2] ) ÷ [key_3]"
+        elements = {'key_1': np.array([...]),   # image
+                    'key_2': np.array([...]),   # image
+                    'key_3': 2}                 # integer
+
+        This will recursively apply (call self.apply_operation method):
+            - M1 = multiplication of key_1 image and 3
+            - M2 = multiplication of key_2 image and 4
+            - A1 = addition of M1 and M2
+            - result = division of A1 and key_3 value
+        """
+        # operations ordered on priorities
+        operations = ['multiply', 'divide', 'add', 'subtract']
+        signs = ['x', '÷', '+', '-']
+
+        # initialize formula
+        if initialize:
+            formula = formula.replace(',', '.')
+            # add space
+            formula = list(formula)
+            for i, f in enumerate(formula):
+                if f in signs + ['(', ')']:
+                    formula[i] = " {} ".format(f)
+                elif f == '[':
+                    formula[i] = " " + f
+                elif f == ']':
+                    formula[i] = f + " "
+            formula = "".join(formula)
+
+            # convert formula to list
+            if isinstance(formula, str):
+                formula = formula.split(' ')
+            while '' in formula:
+                formula.remove('')
```

Comment:
Because of this if statement, formula goes from string to list, according to `initialize`, value, so in the rest of your method, you need to handle both cases. Is it what you wanted to do?

Reply:
no in fact the method must handle only string, I removed this statement